domains/pallets/auto-id/src/tests.rs

```diff
@@ -0,0 +1,228 @@
+use crate::pallet::{AutoIds, NextAutoIdIdentifier};
+use crate::{
+    self as pallet_auto_id, Identifier, Pallet, RegisterAutoId, RegisterAutoIdX509, Signature,
+};
+use codec::Encode;
+use frame_support::dispatch::RawOrigin;
+use frame_support::traits::{ConstU16, ConstU32, ConstU64, Time};
+use pem::parse;
+use ring::rand::SystemRandom;
+use ring::signature::RsaKeyPair;
+use sp_certificate_registry::DerVec;
+use sp_core::H256;
+use sp_runtime::traits::{BlakeTwo256, IdentityLookup};
+use sp_runtime::BuildStorage;
+use std::sync::Arc;
+use subspace_runtime_primitives::Moment;
+use x509_parser::der_parser::asn1_rs::ToDer;
+use x509_parser::oid_registry::OID_PKCS1_SHA256WITHRSA;
+use x509_parser::prelude::{AlgorithmIdentifier, FromDer};
+
+type Block = frame_system::mocking::MockBlock<Test>;
+
+frame_support::construct_runtime!(
+    pub struct Test {
+        System: frame_system,
+        AutoId: pallet_auto_id,
+    }
+);
+
+pub struct MockTime;
+impl Time for MockTime {
+    type Moment = Moment;
+
+    fn now() -> Self::Moment {
+        // valid block time for testing certs
+        1_711_367_658_200
+    }
+}
+
+impl pallet_auto_id::Config for Test {
+    type RuntimeEvent = RuntimeEvent;
+    type Time = MockTime;
+}
+
+impl frame_system::Config for Test {
+    type BaseCallFilter = frame_support::traits::Everything;
+    type BlockWeights = ();
+    type BlockLength = ();
+    type DbWeight = ();
+    type RuntimeOrigin = RuntimeOrigin;
+    type RuntimeCall = RuntimeCall;
+    type RuntimeTask = RuntimeTask;
+    type Nonce = u64;
+    type Hash = H256;
+    type Hashing = BlakeTwo256;
+    type AccountId = u64;
+    type Lookup = IdentityLookup<Self::AccountId>;
+    type Block = Block;
+    type RuntimeEvent = RuntimeEvent;
+    type BlockHashCount = ConstU64<250>;
+    type Version = ();
+    type PalletInfo = PalletInfo;
+    type AccountData = ();
+    type OnNewAccount = ();
+    type OnKilledAccount = ();
+    type SystemWeightInfo = ();
+    type SS58Prefix = ConstU16<42>;
+    type OnSetCode = ();
+    type MaxConsumers = ConstU32<16>;
+}
+
+pub(crate) fn new_test_ext() -> sp_io::TestExternalities {
+    let t = frame_system::GenesisConfig::<Test>::default()
+        .build_storage()
+        .unwrap();
+
+    let mut ext: sp_io::TestExternalities = t.into();
+    ext.register_extension(
+        sp_certificate_registry::host_functions::HostFunctionExtension::new(Arc::new(
+            sp_certificate_registry::host_functions::HostFunctionsImpl,
+        )),
+    );
+    ext
+}
+
+/// Converts Algorithm identifier to Der since x509 does not implement the ToDer :(.
+fn algorithm_to_der(algorithm_identifier: AlgorithmIdentifier) -> DerVec {
```

Comment:
How much do we gain by encoding the algorithm to DER? This adds a layer of complexity in building the SDK so I just want to understand the full benefits.

Reply:
Never mind, wasn't complex.

Reply:
This is done in tests since the specific crate we use did not implment DerEncoding for this specific type but inner types are implemented.

This will not be complex task since python or other DerParsers should be able to implement this automatically